Add XMC47 relax kit  board

talih0 opened the pull request.

talih0 committed:
soc: arm: infineon_xmc: Add XMC4700 MCU series

Adds XMC4700 MCU series.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

talih0 committed:
boards: arm: Add xmc47_relax_kit

Adds Infineon xmc47_relax_kit board.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

talih0 committed:
dts: arm: infineon: xmc4xxx: Define memory regions at each MCU derivative

XMC4500 and XMC47/800 MCUs have a different memory layout. The
definitions have been moved to the derivative .dtsi of each MCU.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

talih0 committed:
tests: drivers: uart_async_api: Add xmc47_relax_kit overlay/conf

Adds overlay and conf files to run the test with the xmc47_relax_kit.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

talih0 committed:
samples: sensor: die_temp_polling: Adds xmc47_relax_kit overlay

Adds overlay to run the sample with the xmc47_relax_kit.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

talih0 committed:
tests: drivers: spi_loopback: Adds xmc47_relax_kit overlay/conf

Adds overlay and conf to run the test on xmc47_relax_kit.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

npal-cy commented on boards/arm/xmc47_relax_kit/xmc47_relax_kit.dts:
```diff
@@ -0,0 +1,109 @@
+		};
+		led2: led2 {
+			gpios = <&gpio5 8 GPIO_ACTIVE_HIGH>;
+		};
```
can we add user button as well?

npal-cy replied:
![image](https://user-images.githubusercontent.com/46566584/227773487-0c4572b4-19ea-401c-86ec-9f7c5bfbb050.png)

talih0 replied:
Unfortunately P15.13, P15.12 don't support interrupts (there is no ERU connection). And all the samples I've seen with sw0 assume interrupt support. Not sure if it's worth adding it. What do you think?

npal-cy replied:
yes, it doesn't make sense to add it.

talih0 commented on dts/arm/infineon/xmc4700_F144x2048.dtsi:
```diff
@@ -0,0 +1,95 @@
+		ngpios = <7>;
+		status = "disabled";
+	};
+};
```
I've used them as analog ports as of now. But you are right, they can be used as a gpio input (no output driver though). I'll update the PR.

talih0 committed:
dts: arm: infineon: xmc4xxx: Add Port 14/15 to device tree

Add Port 14/15 to device tree. These ports can only be configured as input.
Error out in gpio driver if user sets them as output.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

talih0 commented on dts/arm/infineon/xmc4700_F144x2048.dtsi:
```diff
@@ -0,0 +1,101 @@
+};
+
+&gpio15 {
+	ngpios = <12>;
```
@npal-cy, good catch. I think your proposed solution is the cleanest way to do it. I'll update the PR in the next couple of days.

talih0 replied:
@npal-cy dts/bindings/gpio-controller.yaml covers how to handle these cases (see gpio-reserved-ranges). I've updated the PR based on the .yaml file.

npal-cy replied:
@talih0 , you still need to update gpio driver to get correct .port_pin_mask. Currently you are using `.common = {.port_pin_mask = GPIO_PORT_PIN_MASK_FROM_DT_INST(index)}};' and GPIO_PORT_PIN_MASK_FROM_DT_INST does't  handle this.

please look at drivers\gpio\gpio_nxp_s32.c, they also had same case.

talih0 replied:
@npal-cy, I think the proper fix is to fix the GPIO_PORT_PIN_MASK_FROM_DT_INST() macro. Otherwise, each driver will have their own implementation:
- gpio_nxp_s32.c is not great because it can only handle one reserved region. 
- gpio_mcux_igpio.c has a runtime solution, but I think it's better to handle this at compile time.

But currently, I don't think it's possible to merge multiple gpio-reserved-ranges into a common reserved_mask at compile time. Initially I thought we could use LISTIFY(LEN, ....), but unfortunately this doesn't work because LEN needs to be literal constant and we need to iterate over half of the array gpio-reserved-ranges array.

Overall, I think fixing this issue should be a separate PR that will be used cleanly by other gpio drivers.

npal-cy replied:
could you please file ticket for GPIO_PORT_PIN_MASK_FROM_DT_INST update?

talih0 committed:
dts: bindings: gpio: gpio-controller: Fix typo

ngpios should be set to max slot number + 1.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

talih0 commented:
I rebased to get a frech CI run

talih0 commented:
@npal-cy, @ifyall, thank you for approving. I'll rebase this PR after https://github.com/zephyrproject-rtos/zephyr/pull/56676 is merged.

talih0 committed:
tests: drivers: adc_api: Add copyright header to xmc45_relax_kit

Add missing copyright header.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

carlescufi commented:
@talih0 please rebase

talih0 commented:
@npal-cy, @ifyall, could you re-ack when you have a moment, please?
(the CI issue is in mcumgr test is unrelated to the PR)
